src/dragonball/src/device_manager/mod.rs

```diff
@@ -515,16 +598,66 @@ impl DeviceManager {
         self.create_legacy_devices(&mut ctx)?;
         self.init_legacy_devices(dmesg_fifo, com1_sock_path, &mut ctx)?;
 
+        #[cfg(feature = "virtio-blk")]
+        self.block_manager
+            .attach_devices(&mut ctx)
+            .map_err(StartMicroVmError::BlockDeviceError)?;
+
+        #[cfg(feature = "virtio-fs")]
+        {
+            let mut fs_manager = self.fs_manager.lock().unwrap();
+            fs_manager
+                .attach_devices(&mut ctx)
+                .map_err(StartMicroVmError::FsDeviceError)?;
+        }
+
+        #[cfg(feature = "virtio-net")]
+        self.virtio_net_manager
+            .attach_devices(&mut ctx)
+            .map_err(StartMicroVmError::VirtioNetDeviceError)?;
+
         #[cfg(feature = "virtio-vsock")]
         self.vsock_manager.attach_devices(&mut ctx)?;
 
+        #[cfg(feature = "virtio-blk")]
+        self.block_manager
+            .generate_kernel_boot_args(kernel_config)
+            .map_err(StartMicroVmError::DeviceManager)?;
         ctx.generate_kernel_boot_args(kernel_config)
-            .map_err(StartMicrovmError::DeviceManager)?;
+            .map_err(StartMicroVmError::DeviceManager)?;
+
+        Ok(())
+    }
 
+    /// Start all registered devices when booting the associated virtual machine.
+    pub fn start_devices(&mut self) -> std::result::Result<(), StartMicroVmError> {
```

Comment:
When will this method be implemented? Is there another open PR for this?

Reply:
This will be implemented in the future stages of vfio support.
I'll create an issue and a TODO here.